# II- Des bibliothèques importantes pour la manupulation  des donnée

In [1]:
import pandas as pd                                                 
import matplotlib.pyplot as plt                                     
import seaborn as sns                                               
import pandas as pd                                                 
import matplotlib.pyplot as plt                                     
import seaborn as sns                                               
from sklearn.preprocessing import StandardScaler                    
import numpy as np
from sklearn.model_selection import cross_val_score, KFold
from sklearn.neighbors import KNeighborsClassifier                               
from sklearn.metrics import accuracy_score, classification_report, roc_auc_score 
from sklearn.metrics import confusion_matrix                                     
from sklearn.metrics import ConfusionMatrixDisplay
import scikitplot as skplt

# III- Importation du jeu des donnée et description des données

               1) Le jeu de donné est repartie en deux parties :

*   Une partie d’entraînement: Conçu pour entraîner un classeur 
*   Une partie test : conçu pour tester la performance du classeur afin pour prédire le niveau de satisfaction d'un nouveau passager.

Dans un premier temps, nous allons étudier la partie entraînement, fichier train, l’étude de ce fichier est crucial pour nettoyer ses données et le rendre en bonne qualité afin d'entrainner un model de prédiction dans des données propres. Nous allons appeler ce fichier et l'affecter à la variable **df1** :

In [2]:
df1 = pd.read_csv("/home/users/etudiant/Documents/Qualité des donnée  2/train.csv", sep = ',' , header = 0 )

         2) Différentes variables:
Dans nos données trois types des variables se présentent, notamment : 
*   Des variables décrivant le service de l'entreprise, variables 
qualitatives nominales c'est à dire des variables qui prennent des valeurs qui appartiennent à un ensemble fini  dans notre cas ils prennent les valeurs de **0 à 5** ( c'est le niveau de satisfaction dont les passagers donnent pour le service de l'entreprise). Ces variables seront utilisées pour faire des analyses statistiques pour comprendre les différences ou les similitudes entre le niveau de satisfaction des passagers sur le service de l'entreprise. Ils s'agit de l'ensemble  affecté à la variable **variable_service** ci-dessous:

In [4]:
variable_service = ['Inflight wifi service', 'Departure/Arrival time convenient',
                    'Ease of Online booking', 'Gate location', 'Food and drink', 
                    'Online boarding', 'Seat comfort', 'Inflight entertainment',
                    'On-board service', 'Leg room service', 'Baggage handling', 
                       'Checkin service', 'Inflight service', 'Cleanliness']

df1[variable_service].astype(object)

Inflight wifi service Departure/Arrival time convenient  \
0                          3                                 4   
1                          3                                 2   
2                          2                                 2   
3                          2                                 5   
4                          3                                 3   
...                      ...                               ...   
103899                     2                                 1   
103900                     4                                 4   
103901                     1                                 1   
103902                     1                                 1   
103903                     1                                 3   

       Ease of Online booking Gate location Food and drink Online boarding  \
0                           3             1              5               3   
1                           3             3              1               3   
2                           2             2              5               5   
3                           5             5              2               2   
4                           3             3              4               5   
...                       ...           ...            ...             ...   
103899                      2             3              2               2   
103900                      4             4              2               4   
103901                      1             3              4               1   
103902                      1             5              1               1   
103903                      3             3              1               1   

       Seat comfort Inflight entertainment On-board service Leg room service  \
0                 5                      5                4                3   
1                 1                      1                1                5   
2                 5                      5                4                3   
3                 2                      2                2                5   
4                 5                      3                3                4   
...             ...                    ...              ...              ...   
103899            2                      2                3                1   
103900            5                      5                5                5   
103901            5                      4                3                2   
103902            1                      1                4                5   
103903            1                      1                1                1   

       Baggage handling Checkin service Inflight service Cleanliness  
0                     4               4                5           5  
1                     3               1                4           1  
2                     4               4                4           5  
3                     3               1                4           2  
4                     4               3                3           3  
...                 ...             ...              ...         ...  
103899                4               2                3           2  
103900                5               5                5           4  
103901                4               5                5           4  
103902                1               5                4           1  
103903                4               4                3           1  

[103904 rows x 14 columns]

* Des variable quantitatives continue décrivant les vols, il s'agit: 

```
variable_vols = ['Flight Distance','Departure Delay in Minutes', 'Arrival Delay in Minutes']
```

In [5]:
variable_vols = ['Flight Distance','Departure Delay in Minutes','Arrival Delay in Minutes']

df1[variable_vols].astype(object)

Flight Distance Departure Delay in Minutes Arrival Delay in Minutes
0                  460                         25                     18.0
1                  235                          1                      6.0
2                 1142                          0                      0.0
3                  562                         11                      9.0
4                  214                          0                      0.0
...                ...                        ...                      ...
103899             192                          3                      0.0
103900            2347                          0                      0.0
103901            1995                          7                     14.0
103902            1000                          0                      0.0
103903            1723                          0                      0.0

[103904 rows x 3 columns]

*   Des variables quantitatives continue décrivant les passagers, il s'agit:
```
variable_passager = ['Gender', 'Customer Type', 'Age','Type of Travel','Class']
```

Une petite réflexe est de savoir combien des données on a dans le fichier **df1** définissant dès maintenant l'ensemble d'entrainement, alors on peut le lire facilement tout en bas de chaque tableau précédent que l'ensemble d'entrainement contient 103904 données dont 25 colonnes, on peut ainsi s'amuser à le voir d'une autre façon (sous forme dataFRame):

In [6]:
df1.shape

(103904, 25)

Donc le jeu des données d’entraînement contient 103904 exemples dont 25 colonnes.
Un exemple est un couple (d,Y) où d est un ensemble des données et Y une classe.
Notre jeu des données contient 2 classe, notamment:

               satisfed et neutral or dissatisfied

Le travail est de nettoyer ce fichier pour préparer un champs d’étude propre afin de mieux prédire chaque donnée sa classe. 

On peut visualiser à quoi ressemble l'ensemble des données:

In [7]:
df1

Unnamed: 0      id  Gender      Customer Type  Age   Type of Travel  \
0                0   70172    Male     Loyal Customer   13  Personal Travel   
1                1    5047    Male  disloyal Customer   25  Business travel   
2                2  110028  Female     Loyal Customer   26  Business travel   
3                3   24026  Female     Loyal Customer   25  Business travel   
4                4  119299    Male     Loyal Customer   61  Business travel   
...            ...     ...     ...                ...  ...              ...   
103899      103899   94171  Female  disloyal Customer   23  Business travel   
103900      103900   73097    Male     Loyal Customer   49  Business travel   
103901      103901   68825    Male  disloyal Customer   30  Business travel   
103902      103902   54173  Female  disloyal Customer   22  Business travel   
103903      103903   62567    Male     Loyal Customer   27  Business travel   

           Class  Flight Distance  Inflight wifi service  \
0       Eco Plus              460                      3   
1       Business              235                      3   
2       Business             1142                      2   
3       Business              562                      2   
4       Business              214                      3   
...          ...              ...                    ...   
103899       Eco              192                      2   
103900  Business             2347                      4   
103901  Business             1995                      1   
103902       Eco             1000                      1   
103903  Business             1723                      1   

        Departure/Arrival time convenient  ...  Inflight entertainment  \
0                                       4  ...                       5   
1                                       2  ...                       1   
2                                       2  ...                       5   
3                                       5  ...                       2   
4                                       3  ...                       3   
...                                   ...  ...                     ...   
103899                                  1  ...                       2   
103900                                  4  ...                       5   
103901                                  1  ...                       4   
103902                                  1  ...                       1   
103903                                  3  ...                       1   

        On-board service  Leg room service  Baggage handling  Checkin service  \
0                      4                 3                 4                4   
1                      1                 5                 3                1   
2                      4                 3                 4                4   
3                      2                 5                 3                1   
4                      3                 4                 4                3   
...                  ...               ...               ...              ...   
103899                 3                 1                 4                2   
103900                 5                 5                 5                5   
103901                 3                 2                 4                5   
103902                 4                 5                 1                5   
103903                 1                 1                 4                4   

        Inflight service  Cleanliness  Departure Delay in Minutes  \
0                      5            5                          25   
1                      4            1                           1   
2                      4            5                           0   
3                      4            2                          11   
4                      3            3                           0   
...                  ...          ...                         

In [8]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 25 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   Unnamed: 0                         103904 non-null  int64  
 1   id                                 103904 non-null  int64  
 2   Gender                             103904 non-null  object 
 3   Customer Type                      103904 non-null  object 
 4   Age                                103904 non-null  int64  
 5   Type of Travel                     103904 non-null  object 
 6   Class                              103904 non-null  object 
 7   Flight Distance                    103904 non-null  int64  
 8   Inflight wifi service              103904 non-null  int64  
 9   Departure/Arrival time convenient  103904 non-null  int64  
 10  Ease of Online booking             103904 non-null  int64  
 11  Gate location                      1039

On a alors des variable de types objet et de type int, ces variables se repartissent de la façon suivant: 
* des variables catégorique, qui représentent les notes données par chaque passager au service de l'entreprise,
* des données quantitative continue, on vas bien les étudier dans la suite.

                    Remarque sur cette visualision

Réduire les variables du jeu des données est utile si l'on peut, puisque ça va facilité la tache surtout sur l’entraînement du modèle de classification. Parmi les méthodes d’éliminations des colonnes, on peut étudier la corrélations des variables pour voir ce qui sont corrélés et les éliminer en avantage, cet étape on le verra dans la suite, pour l'instant on s'en occupe de quelque colonnes qui nous saute aux yeux:
- On constate que la 1er colonne **Unnamed 0** ne servira pas pour l'analyse 
statistique, par défaut python affecte un 'id' à chacun des passagers, vaut mieux le supprimer.
- De même la colonnes **id** ne servira à rien parce puisqu'on est pas obligé de savoir l'id du passager pour définir sa niveau de satisfaction:

In [9]:
df_clean = df1.drop(['Unnamed: 0', 'id'], axis = 1)
df_clean

Gender      Customer Type  Age   Type of Travel     Class  \
0         Male     Loyal Customer   13  Personal Travel  Eco Plus   
1         Male  disloyal Customer   25  Business travel  Business   
2       Female     Loyal Customer   26  Business travel  Business   
3       Female     Loyal Customer   25  Business travel  Business   
4         Male     Loyal Customer   61  Business travel  Business   
...        ...                ...  ...              ...       ...   
103899  Female  disloyal Customer   23  Business travel       Eco   
103900    Male     Loyal Customer   49  Business travel  Business   
103901    Male  disloyal Customer   30  Business travel  Business   
103902  Female  disloyal Customer   22  Business travel       Eco   
103903    Male     Loyal Customer   27  Business travel  Business   

        Flight Distance  Inflight wifi service  \
0                   460                      3   
1                   235                      3   
2                  1142                      2   
3                   562                      2   
4                   214                      3   
...                 ...                    ...   
103899              192                      2   
103900             2347                      4   
103901             1995                      1   
103902             1000                      1   
103903             1723                      1   

        Departure/Arrival time convenient  Ease of Online booking  \
0                                       4                       3   
1                                       2                       3   
2                                       2                       2   
3                                       5                       5   
4                                       3                       3   
...                                   ...                     ...   
103899                                  1                       2   
103900                                  4                       4   
103901                                  1                       1   
103902                                  1                       1   
103903                                  3                       3   

        Gate location  ...  Inflight entertainment  On-board service  \
0                   1  ...                       5                 4   
1                   3  ...                       1                 1   
2                   2  ...                       5                 4   
3                   5  ...                       2                 2   
4                   3  ...                       3                 3   
...               ...  ...                     ...               ...   
103899              3  ...                       2                 3   
103900              4  ...                       5                 5   
103901              3  ...                       4                 3   
103902              5  ...                       1                 4   
103903              3  ...                       1                 1   

        Leg room service  Baggage handling  Checkin service  Inflight service  \
0                      3                 4                4                 5   
1                      5                 3                1                 4   
2                      3                 4                4                 4   
3                      5                 3                1                 4   
4                      4                 4                3                 3   
...                  ...               ...              ...               ...   
103899                 1                 4                2                 3   
103900                 5                 5                5                 5   
103901                 2                 4                5                 5   
103902                 5                 1                5               

Donc jusqu'à là, le jeu des données est réduit à 23 colonnes avec les 103904 lignes.


               3) Division des données :
Dans nos données il y a deux différentes variables : 

*  Des **variables qualitatives nominales**, c'est à dire des variables à valeur dans un ensemble fini, ce sont les variables catégoriques qui prennent des valeurs dans **[0,5]**. Pour chacun de service de l'entreprise ces valeurs définissent le niveau de satisfaction de chaque passagers enregistré. De même des variables catégoriques, ils s'agit **Gender,	Customer Type,  Type of Travel, Class** qui seront transformé en variable quantitative continue  à valeurs dans un ensemble fini, on verra la manipulation dans la suite. Ces variables seront utilisées pour faire des analyses statistiques pour comprendre les différences ou les similitudes entre les groupes de données. Les noms de ces variables sont consignés  **variable_nominale** ci-dessous.

        variable_nominale = ['Gender', 'Customer Type', 'Type of Travel', 'Class', 'Inflight wifi service', 
                              'Departure/Arrival time convenient', 'Ease of Online booking', 'Gate location',
                              'Food and drink', 'Online boarding', 'Seat comfort', 'Inflight entertainment', 
                              'On-board service', 'Leg room service', 'Baggage handling', 'Checkin service', 
                              'Inflight service', 'Cleanliness', 'satisfaction']
                              
*   Des **variables quantitatives continues**: il s'agit l'ensemble des variables dans **variable_continue** ci-dessous. Ces variables continuent la description de chaque passager.
 
         variable_continue = ['Age', 'Flight Distance', 'Departure Delay in Minutes', 'Arrival Delay in Minutes']

         

In [10]:
variable_nominale = ['Gender', 'Customer Type', 'Type of Travel', 'Class', 'Inflight wifi service', 
                       'Departure/Arrival time convenient', 'Ease of Online booking', 'Gate location',
                       'Food and drink', 'Online boarding', 'Seat comfort', 'Inflight entertainment', 
                       'On-board service', 'Leg room service', 'Baggage handling', 'Checkin service', 
                       'Inflight service', 'Cleanliness', 'satisfaction']

variable_continue = ['Age', 'Flight Distance', 'Departure Delay in Minutes', 'Arrival Delay in Minutes']

In [11]:
df_clean[variable_nominale].astype(object)

Gender      Customer Type   Type of Travel     Class  \
0         Male     Loyal Customer  Personal Travel  Eco Plus   
1         Male  disloyal Customer  Business travel  Business   
2       Female     Loyal Customer  Business travel  Business   
3       Female     Loyal Customer  Business travel  Business   
4         Male     Loyal Customer  Business travel  Business   
...        ...                ...              ...       ...   
103899  Female  disloyal Customer  Business travel       Eco   
103900    Male     Loyal Customer  Business travel  Business   
103901    Male  disloyal Customer  Business travel  Business   
103902  Female  disloyal Customer  Business travel       Eco   
103903    Male     Loyal Customer  Business travel  Business   

       Inflight wifi service Departure/Arrival time convenient  \
0                          3                                 4   
1                          3                                 2   
2                          2                                 2   
3                          2                                 5   
4                          3                                 3   
...                      ...                               ...   
103899                     2                                 1   
103900                     4                                 4   
103901                     1                                 1   
103902                     1                                 1   
103903                     1                                 3   

       Ease of Online booking Gate location Food and drink Online boarding  \
0                           3             1              5               3   
1                           3             3              1               3   
2                           2             2              5               5   
3                           5             5              2               2   
4                           3             3              4               5   
...                       ...           ...            ...             ...   
103899                      2             3              2               2   
103900                      4             4              2               4   
103901                      1             3              4               1   
103902                      1             5              1               1   
103903                      3             3              1               1   

       Seat comfort Inflight entertainment On-board service Leg room service  \
0                 5                      5                4                3   
1                 1                      1                1                5   
2                 5                      5                4                3   
3                 2                      2                2                5   
4                 5                      3                3                4   
...             ...                    ...              ...              ...   
103899            2                      2                3                1   
103900            5                      5                5                5   
103901            5                      4                3                2   
103902            1                      1                4                5   
103903            1                      1                1                1   

       Baggage handling Checkin service Inflight service Cleanliness  \
0                     4               4                5           5   
1                     3               1                4           1   
2                     4               4                4           5   
3                     3               1                4           2   
4                     4               3                3           3   
...                 ...             ...              ...         ...   
103899                4               2     

In [12]:
df_clean[variable_continue].astype(object)

Age Flight Distance Departure Delay in Minutes Arrival Delay in Minutes
0       13             460                         25                     18.0
1       25             235                          1                      6.0
2       26            1142                          0                      0.0
3       25             562                         11                      9.0
4       61             214                          0                      0.0
...     ..             ...                        ...                      ...
103899  23             192                          3                      0.0
103900  49            2347                          0                      0.0
103901  30            1995                          7                     14.0
103902  22            1000                          0                      0.0
103903  27            1723                          0                      0.0

[103904 rows x 4 columns]